src/humancellatlas/data/metadata/helpers/staging_area.py

```diff
@@ -219,117 +207,103 @@ def _entity_ids_by_type(self,
-        path = () if path == '' else tuple(path.split('/'))
-        return cls(repo=repo, ref=ref, path=path)
+        log.debug("Cloning '%s' into '%s'", url, tmpdir)
+        repo = git.Repo.clone_from(url, tmpdir)
+        commit = repo.commit(ref)
+        tree = repo.tree(commit)
+        tree = cls._get_subtree(tree, path)
+        return cls(url=url, repo=repo, commit=commit, path=path, tree=tree)
 
     @classmethod
-    def from_url(cls, url: str) -> 'GitHubStagingAreaFactory':
-        """
-        :param url: The URL of a staging area in a GitHub repository with syntax
-                    `https://github.com/<OWNER>/<NAME>/tree/<REF>[/<PATH>]`.
-                    Note that REF can be a branch, tag, or commit SHA. If REF
-                    contains special characters like `/`, '?` or `#` they must
-                    be URL-encoded. This is especially noteworthy for `/` since
-                    it's the only way to distinguish slashes in REF from those
-                    in PATH. The slashes in PATH must not be URL-encoded, while
-                    occurrences of `#` and `?` must.
-        """
-        parsed_url = furl(url)
-        require(parsed_url.scheme == 'https', url)
-        require(parsed_url.host == 'github.com', url)
-        require(len(parsed_url.path.segments) > 3, url)
-        require(parsed_url.path.segments[2] == 'tree', url)
-        owner, name = parsed_url.path.segments[0:2]
-        ref = parsed_url.path.segments[3]
-        path = '/'.join(parsed_url.path.segments[4:])
-        return cls.create(owner=owner, name=name, ref=ref, path=path)
+    def _get_subtree(cls, root: git.Tree, path: tuple[str, ...]) -> git.Tree:
+        while len(path) > 0:
+            p, *path = path
+            for tree in root.trees:
+                if tree.name == p:
+                    root = tree
+                    break
+            else:
+                raise ValueError(f'Could not find path {p!r} in {root.path!r}')
+        return root
 
     def load_staging_area(self) -> StagingArea:
-        staging_area_folders = self._get_folders(path=self.path)
+        staging_area_folders = {t.name for t in self.tree.trees}
         expected_folders = {'data', 'descriptors', 'links', 'metadata'}
         require(set(staging_area_folders) == expected_folders,
-                f'{self.repo.full_name} {self.path} is not a valid staging area')
-        links = self._get_files(path=self.path + ('links',))
-        metadata = {}
-        for folder in self._get_folders(path=self.path + ('metadata',)):
-            files = self._get_files(path=self.path + ('metadata', folder))
-            metadata.update(files)
-        descriptors = {}
-        for folder in self._get_folders(path=self.path + ('descriptors',)):
-            files = self._get_files(path=self.path + ('descriptors', folder))
-            descriptors.update(files)
+                f'{self.url} {self.tree.path} is not a valid staging area')
```

Comment:
The string composition here doesn't make syntactic or grammatical sense to me. I would just write
```suggestion
                'Invalid staging area', self.url, self.tree.path)
```
but maybe there's a more elegant way of correctly combining the url and path?